Add SOFT_ASSERT to gracefully recover from invariant violations

davidchencsl opened the pull request:
Summary: Implement SOFT_ASSERT that only fails in debug mode, but only trigger a warning log in release mode. This allows us to gracefully handle some of the invariant violation when processing traces that doesn't necessarily need to crash the entire program.

Test Plan: Added SOFT_ASSERT test in containers.cpp

Differential Revision: D38327334

robieta commented on torch/csrc/profiler/util.h:
```diff
@@ -35,6 +35,21 @@
 
+// Define SOFT_ASSERT
+#ifdef NDEBUG
+#define HANDLE_ERROR(cond, ...) TORCH_WARN(__VA_ARGS__);
```
This should be runtime toggle-able, no?

robieta commented on torch/csrc/profiler/util.h:
```diff
@@ -35,6 +35,24 @@
 #endif
 
+// Define SOFT_ASSERT
+#ifdef NDEBUG
```
As written, `NDEBUG` takes priority over the environment variable. I think it should be the other way around. (And of course that means the environment variable must be able to both force enable and force disable.) My reasoning is that I want to add unit test for the profiler that test various kinds of pathological conditions / invalid inputs and assert that they are handled gracefully. However if the behavior depends on the build flag then it's impossible to write unit tests that will always work.

davidchencsl replied:
Maybe let the env var disable the HANDLE_ERROR in both cases?

davidchencsl replied:
Ah I see what you are saying.

albanD reviewed (changes requested):
I think this should be discussed in the context of https://github.com/pytorch/rfcs/pull/43 before adding a new API.

In particular, you could be raising a warning of a custom subclass and then use python warning filter to make this warning class an error if that's what you want for particular runs!

robieta commented on torch/csrc/profiler/util.h:
```diff
@@ -48,9 +65,9 @@ using steady_clock_t = std::conditional<
   auto now = std::chrono::system_clock::now().time_since_epoch();
   return std::chrono::duration_cast<std::chrono::nanoseconds>(now).count();
-}
+  }
```
nit: unrelated formatting?

robieta commented on torch/csrc/profiler/util.h:
```diff
@@ -35,9 +36,25 @@
 
+// Define SOFT_ASSERT
+#define SOFT_ASSERT(cond, ...)                                     \
+  [&](const char* func, const char* file, uint32_t line) -> bool { \
```
We should use these, since we went through the trouble of plumbing them in.

robieta reviewed (commented):
Only major comment is to include the source info. (Tip: `c10::str` will help you package it, since you can't use fmt in a header.)

robieta commented:
> I think this should be discussed in the context of [pytorch/rfcs#43](https://github.com/pytorch/rfcs/pull/43) before adding a new API.
> 
> In particular, you could be raising a warning of a custom subclass and then use python warning filter to make this warning class an error if that's what you want for particular runs!

Oh, interesting. I was not aware of this but it seems like exactly what we were going for. (Basically a warn that fails unit tests.)

robieta commented:
@albanD What are your thoughts on landing this with the understanding that it's just a placeholder until we can move to https://github.com/pytorch/rfcs/pull/43. Because as awesome as that sounds, we do have a concrete need today. (And we can hide the APIs even more if you're concerned that they're not private enough.)

albanD commented:
We plan on moving quite quickly on this, especially the part related to error and warning.

I think the only thing that is missing for you to be able to use the normal warning filter is for you to be able to warn with a custom subclass on the c++ side.
cc @kurtamohler that might be a good first step (c++ warning subclass and c++ warning filter) for the rfc implementation since we already have a need here haha

If you want to merge this before the warning is done, could you:
- Add a comment that this API will be changed soon once the rfc is done.
- Do you actually want to keep different behavior for debug vs non-debug build? I think this is the only thing that would prevent you from moving the API introduced here to use a warning-based system once it is available.

Impl sounds good to me otherwise.

robieta commented:
> Add a comment that this API will be changed soon once the rfc is done.

Absolutely. This looks absolutely awesome and I'd be happy to work with @kurtamohler as a concrete early use case.

> Do you actually want to keep different behavior for debug vs non-debug build?

I would argue that an analog of `TORCH_INTERNAL_ASSERT_DEBUG_ONLY` would be useful. However it's not essential, and if it would slow things down I think we should drop it and just do what the rfc implementation can support. Better to move quickly to the idiomatic way ASAP and build from there.

robieta commented on torch/csrc/profiler/util.h:
```diff
@@ -35,9 +36,25 @@
 #endif
 #endif
 
+// Define SOFT_ASSERT
```
// TODO: replace with https://github.com/pytorch/rfcs/pull/43 when it is ready.

robieta reviewed (approved):
LGTM. (Modulo Alban's request for a comment.)

albanD commented:
@robieta quick question: Do you need the filtering system that can transform warnings into error to work in a c++-only world? Or is that ok if that is only available from python?

robieta commented:
> @robieta quick question: Do you need the filtering system that can transform warnings into error to work in a c++-only world? Or is that ok if that is only available from python?

It's not ideal (we want to write more scoped C++ unit tests for C++ components), but not the end of the world.

davidchencsl committed:
Add SOFT_ASSERT to gracefully recover from invariant violations (#82689)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/82689

Implement SOFT_ASSERT that only fails in debug mode, but only trigger a warning log in release mode. This allows us to gracefully handle some of the invariant violation when processing traces that doesn't necessarily need to crash the entire program.

Test Plan: Added SOFT_ASSERT test in containers.cpp

Reviewed By: robieta

Differential Revision: D38327334

fbshipit-source-id: 358c21cf23a0c6c66070e4ffe28340a440fe5421

davidchencsl commented:
@pytorchbot merge -g
